File: python/tests/test_pathfinding.py
import unittest

from snake_game.ai.pathfinding import astar_path, bfs_path, direction_between, flood_fill_count


class BfsPathTests(unittest.TestCase):
    def test_finds_the_shortest_path_on_an_open_grid(self):
        path = bfs_path((0, 0), (2, 0), blocked=set(), cols=5, rows=5)
        self.assertEqual(path, [(0, 0), (1, 0), (2, 0)])

    def test_routes_around_obstacles(self):
        blocked = {(1, 0), (1, 1)}
        path = bfs_path((0, 0), (2, 0), blocked, cols=3, rows=3)
        self.assertIsNotNone(path)
        self.assertTrue(all(cell not in blocked for cell in path))

    def test_returns_none_when_the_goal_is_walled_off(self):
        blocked = {(1, y) for y in range(5)}
        path = bfs_path((0, 0), (4, 0), blocked, cols=5, rows=5)
        self.assertIsNone(path)


class AStarPathTests(unittest.TestCase):
    def test_matches_bfs_path_length_on_an_open_grid(self):
        bfs = bfs_path((0, 0), (4, 4), set(), cols=5, rows=5)
        astar = astar_path((0, 0), (4, 4), set(), cols=5, rows=5)
        self.assertEqual(len(bfs), len(astar))

    def test_returns_none_when_unreachable(self):
        blocked = {(1, y) for y in range(5)}
        self.assertIsNone(astar_path((0, 0), (4, 0), blocked, cols=5, rows=5))


class FloodFillTests(unittest.TestCase):
    def test_counts_an_enclosed_region_correctly(self):
        # wall at col 2 splits 5x5 grid into two pockets
        blocked = {(2, y) for y in range(5)}
        count = flood_fill_count((0, 0), blocked, cols=5, rows=5)
        self.assertEqual(count, 10)

    def test_returns_zero_when_start_is_blocked(self):
        self.assertEqual(flood_fill_count((0, 0), blocked={(0, 0)}, cols=5, rows=5), 0)


class DirectionBetweenTests(unittest.TestCase):
    def test_returns_the_unit_step_between_adjacent_cells(self):
        self.assertEqual(direction_between((3, 3), (4, 3)), (1, 0))
        self.assertEqual(direction_between((3, 3), (3, 2)), (0, -1))


if __name__ == '__main__':
    unittest.main()

File: cpp/tests/test_pathfinding.cpp
#include "ai/pathfinding.hpp"
#include "doctest.h"

using pathfinding::AStarPath;
using pathfinding::BfsPath;
using pathfinding::DirectionBetween;
using pathfinding::FloodFillCount;

TEST_CASE("bfs finds the shortest path on an open grid") {
    auto path = BfsPath(Cell{0, 0}, Cell{2, 0}, {}, 5, 5);
    REQUIRE(path.has_value());
    std::vector<Cell> expected{Cell{0, 0}, Cell{1, 0}, Cell{2, 0}};
    CHECK(*path == expected);
}

TEST_CASE("bfs routes around obstacles") {
    std::unordered_set<Cell, CellHash> blocked{Cell{1, 0}, Cell{1, 1}};
    auto path = BfsPath(Cell{0, 0}, Cell{2, 0}, blocked, 3, 3);
    REQUIRE(path.has_value());
    for (const Cell& c : *path) CHECK(blocked.count(c) == 0);
}

TEST_CASE("bfs returns nothing when the goal is walled off") {
    std::unordered_set<Cell, CellHash> blocked;
    for (int y = 0; y < 5; ++y) blocked.insert(Cell{1, y});
    CHECK_FALSE(BfsPath(Cell{0, 0}, Cell{4, 0}, blocked, 5, 5).has_value());
}

TEST_CASE("astar matches bfs path length on an open grid") {
    auto bfs = BfsPath(Cell{0, 0}, Cell{4, 4}, {}, 5, 5);
    auto astar = AStarPath(Cell{0, 0}, Cell{4, 4}, {}, 5, 5);
    REQUIRE(bfs.has_value());
    REQUIRE(astar.has_value());
    CHECK(bfs->size() == astar->size());
}

TEST_CASE("astar returns nothing when unreachable") {
    std::unordered_set<Cell, CellHash> blocked;
    for (int y = 0; y < 5; ++y) blocked.insert(Cell{1, y});
    CHECK_FALSE(AStarPath(Cell{0, 0}, Cell{4, 0}, blocked, 5, 5).has_value());
}

TEST_CASE("flood fill counts an enclosed region correctly") {
    // A wall at col 2 splits a 5x5 grid into two 2-wide pockets.
    std::unordered_set<Cell, CellHash> blocked;
    for (int y = 0; y < 5; ++y) blocked.insert(Cell{2, y});
    CHECK(FloodFillCount(Cell{0, 0}, blocked, 5, 5) == 10);
}

TEST_CASE("flood fill returns zero when start is blocked") {
    CHECK(FloodFillCount(Cell{0, 0}, {Cell{0, 0}}, 5, 5) == 0);
}

TEST_CASE("direction between adjacent cells") {
    CHECK(DirectionBetween(Cell{3, 3}, Cell{4, 3}) == Direction::Right);
    CHECK(DirectionBetween(Cell{3, 3}, Cell{3, 2}) == Direction::Up);
}
